sky/backends/task_codegen.py

```diff
@@ -0,0 +1,633 @@
+                    msg += f'. See error logs above for more details.{colorama.Style.RESET_ALL}'
+                    print(msg, flush=True)
+                    job_lib.set_status({self.job_id!r}, job_lib.JobStatus.FAILED_SETUP)
+                    # This waits for all streaming logs to finish.
+                    time.sleep(1)
+                    # Need this to set the job status in ray job to be FAILED.
+                    sys.exit(1)
+                """)
+            ]
+
+        self._code.append(f'job_lib.set_job_started({self.job_id!r})')
+        if setup_cmd is None:
+            # Need to call schedule_step() to make sure the scheduler
+            # schedule the next pending job.
+            self._code.append('job_lib.scheduler.schedule_step()')
+
+        # Export IP and node rank to the environment variables.
+        self._code += [
+            textwrap.dedent(f"""\
+                @ray.remote
+                def check_ip():
+                    return ray.util.get_node_ip_address()
+                gang_scheduling_id_to_ip = ray.get([
+                    check_ip.options(
+                            num_cpus={task_cpu_demand},
+                            scheduling_strategy=ray.util.scheduling_strategies.PlacementGroupSchedulingStrategy(
+                                placement_group=pg,
+                                placement_group_bundle_index=i
+                            )).remote()
+                    for i in range(pg.bundle_count)
+                ])
+
+                cluster_ips_to_node_id = {{ip: i for i, ip in enumerate({stable_cluster_internal_ips!r})}}
+                job_ip_rank_list = sorted(gang_scheduling_id_to_ip, key=cluster_ips_to_node_id.get)
+                job_ip_rank_map = {{ip: i for i, ip in enumerate(job_ip_rank_list)}}
+                job_ip_list_str = '\\n'.join(job_ip_rank_list)
+                """),
+        ]
+
+    def add_tasks(self,
+                  num_nodes: int,
+                  bash_script: Optional[str],
+                  task_name: Optional[str],
+                  resources_dict: Dict[str, float],
+                  log_dir: str,
+                  env_vars: Optional[Dict[str, str]] = None) -> None:
+        # TODO(zhwu): The resources limitation for multi-node ray.tune and
+        # horovod should be considered.
+        for i in range(num_nodes):
+            # Ray's per-node resources, to constrain scheduling each command to
+            # the corresponding node, represented by private IPs.
+            self.add_single_task(bash_script=bash_script,
+                                 task_name=task_name,
+                                 resources_dict=resources_dict.copy(),
+                                 log_dir=log_dir,
+                                 env_vars=env_vars,
+                                 gang_scheduling_id=i)
+
+    def add_single_task(self,
```

Comment:
```suggestion
    def _add_single_task(self,
```
unless in the following slurm implementation you mean this function to be accessible from outside the package.